guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc

```diff
@@ -0,0 +1,188 @@
+
+source:ProductRepository[app=testcontainers]
+
+callout:repository[1]
+callout:jparepository[2]
+callout:at-query[3]
+
+== Create the event payload java bean
+
+Let us create a domain object named ProductPriceChangedEvent as a record representing the structure of the event payload
+that we are going to receive from the Kafka topic.
+
+source:ProductPriceChangedEvent[app=testcontainers]
+
+callout:serdeable[1]
+
+== Implement Kafka Listener
+
+Finally, let us implement the Kafka listener which handles the messages received from the *product-price-changes* topic
+and updates the product price in the database.
+
+common:kafka-listener[]
+
+source:ProductPriceChangedEventHandler[app=testcontainers]
+
+callout:singleton[1]
+callout:transactional[2]
+callout:constructor-di[number=3,arg0=ProductRepository]
+callout:kafka-topic-listener[4]
+callout:kafka-listener-offset-reset-earliest[5]
+
+Let us assume that there is an agreement between the sender and receiver that the payload will be sent in the following JSON format:
+
+[source,json]
+----
+{
+    "productCode": "P100",
+    "price": 25.00
+}
+----
+
+== Testing
+
+=== Write Test for Kafka Listener
+We are going to write a test for the Kafka event listener *ProductPriceChangedEventHandler*
+by sending a message to the product-price-changes topic and verify the updated product price in the database.
+
+But in order to successfully start our Micronaut context we need Kafka and the MySQL database up and running
+and configure the Micronaut context to talk to them.
+
+Create a `@KafkaClient` to simplify publishing events in the test.
+
+test:ProductPriceChangesClient[app=testcontainers]
+
+callout:kafka-client.adoc[1]
+callout:kafka-topic-publisher.adoc[2]
+callout:kafka-key.adoc[3]
+
+We will use the Testcontainers library to spin up a Kafka and the MySQL database instances as Docker containers
+and configure the application to talk to them as follows:
+
+test:ProductPriceChangedEventHandlerTest[app=testcontainers]
+
+callout:micronaut-test-transactional-false[1]
+<2> Annotate the class with `@Property` to supply the driver class name configuration to the test.
+<3> We have configured the https://java.testcontainers.org/modules/databases/jdbc/[Testcontainers special JDBC URL] to spin up a MySQL container and configure it as a DataSource with Micronaut application context.
+callout:test-containers-disabled-without-docker[4]
+callout:test-instance-per-class[5]
+callout:test-property-provider[6]
+<7> We have used the Testcontainers JUnit 5 Extension annotations *@Testcontainers* and *@Container* to spin up a Kafka container and registered the *bootstrap-servers* location using *TestPropertyProvider* mechanism.
+<8> We have created a Product record in the database before sending the event.
+<9> During the test, we sent a message to the *product-price-changes* topic using *KafkaTemplate* with productCode as key and *ProductPriceChangedEvent* instance as value.
```

Comment:
Update `KafkaTemplate` to `ProductPriceChangesClient`.

Reply:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/a953961322ba0f99e4c29d78b52456a50b44fef3